### PR TITLE
[Ant-2380] improving maintainability

### PR DESCRIPTION
**Pull request by a-zakir**
This pull request includes several changes to the `BendersBase` class and related components to improve the handling of criterion computations and refactor the codebase. The most important changes include the renaming of `OuterLoopInputData` to `CriterionInputData`, the addition of the `CouplingMapGenerator` class, and the refactoring of methods related to criterion computations.

### Refactoring and Renaming:
* Renamed `OuterLoopInputData` to `CriterionInputData` and updated all related references in `CriterionComputation.cpp` and `CriterionInputDataReader.cpp` [[1]](diffhunk://#diff-d0e3218aceff8618ccc9bf1b8ee0ed104175f6db50b089e0e2fd46c4372daf88L32-R44) [[2]](diffhunk://#diff-68cda14d5a160cbb0d388b11de54f391a6bf6aa2a2b46ee4f97dba03c963fedcL1-R1) [[3]](diffhunk://#diff-68cda14d5a160cbb0d388b11de54f391a6bf6aa2a2b46ee4f97dba03c963fedcL13-R83).
* Removed the `OuterloopOptionsFile` method from `BendersBase` and updated the `setCriterionsComputation` method to `setCriterionComputationInputs` [[1]](diffhunk://#diff-debcbcbc301c4e3ac700146a75db0a05d686e1d30f20ce42019e1e77acf903a5L25-L29) [[2]](diffhunk://#diff-36a63705b7409199b9d45a79b1f80f656b02e9f5a4d1a2c41703742ad41a76c2L98-R103) [[3]](diffhunk://#diff-36a63705b7409199b9d45a79b1f80f656b02e9f5a4d1a2c41703742ad41a76c2L226-R225).

### Addition of CouplingMapGenerator:
* Added a new `CouplingMapGenerator` class to handle the building of input maps from structure files, encapsulating this functionality [[1]](diffhunk://#diff-e5c5805ed18be849a2bcfcaf5f3a459394022162fe8c0b168ad0164688aac93eR1-R52) [[2]](diffhunk://#diff-2a07bb398e3d2d00e5899c2b0c02ae4da892bdcf20158651d1a4ce4c1a7b49c3R1-R9).

### Updates to Criterion Computation:
* Refactored `CriterionComputation` to use `CriterionInputData` instead of `OuterLoopInputData`, updating methods and member variables accordingly [[1]](diffhunk://#diff-d0e3218aceff8618ccc9bf1b8ee0ed104175f6db50b089e0e2fd46c4372daf88L32-R44) [[2]](diffhunk://#diff-debcbcbc301c4e3ac700146a75db0a05d686e1d30f20ce42019e1e77acf903a5L420-R415) [[3]](diffhunk://#diff-debcbcbc301c4e3ac700146a75db0a05d686e1d30f20ce42019e1e77acf903a5R1045-R1054).

### CMakeLists and File Updates:
* Updated `CMakeLists.txt` to reflect the renaming of files and the addition of new source files (`CouplingMapGenerator.cpp`) [[1]](diffhunk://#diff-89a10ccfdde5ce80ade12658a06704594fe58b1f86f162a55bc68252f382ff09L26-R28) [[2]](diffhunk://#diff-89a10ccfdde5ce80ade12658a06704594fe58b1f86f162a55bc68252f382ff09L44-L50).

### Removal of Unused Methods:
* Removed the `build_input` method from `common.cpp` as its functionality is now encapsulated in `CouplingMapGenerator`.

### CODE REVIEW

File: src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/CriterionComputation.h [hunk 1]
```diff
@@ -69,10 +74,10 @@ class CriterionComputation {
    *
    * @return A constant reference to the OuterLoopInputData object.
    */
-  const OuterLoopInputData &getOuterLoopInputData() const;
+  const CriterionInputData &getOuterLoopInputData() const;
 
  private:
-  std::vector<std::vector<int>> var_indices_;
-  const OuterLoopInputData outer_loop_input_data_;
```
Comment (tbittar) [inline src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/CriterionComputation.h:76]: Could this attribute also be renamed `criterion_input_data` for consistance ?

File: src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/CriterionComputation.h
Comment (tbittar) [inline src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/CriterionComputation.h:1]: - Rename `ComputeOuterLoopCriterion` as `ComputeCriterion`
- Remove all outer loop references in this function variables
- Remove references to outer loops in the comments and descriptions of functions

File: src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/VariablesGroup.h [hunk 2]
```diff
@@ -3,18 +3,19 @@
...
 
-#include "OuterLoopInputDataReader.h"
+#include "CriterionInputDataReader.h"
 namespace Benders::Criterion {
 class VariablesGroup {
  public:
   explicit VariablesGroup(const std::vector<std::string>& all_variables,
-                          const std::vector<OuterLoopSingleInputData>& outer_loop_single_input_data);
+                          const std::vector<CriterionSingleInputData>&
+                              outer_loop_single_input_data);
```
Comment (tbittar) [inline src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/VariablesGroup.h]:
```suggestion
                              criterion_single_input_data);
```

File: src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/VariablesGroup.h [hunk 3]
```diff
@@ -3,18 +3,19 @@
...
   [[nodiscard]] std::vector<std::vector<int>> Indices() const;
 
  private:
   void Search();
   const std::vector<std::string>& all_variables_;
-  const std::vector<OuterLoopSingleInputData>& outer_loop_single_input_data_;
+  const std::vector<CriterionSingleInputData>& outer_loop_single_input_data_;
```
Comment (tbittar) [inline src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/VariablesGroup.h]:
```suggestion
  const std::vector<CriterionSingleInputData>& criterion_single_input_data_;
```

File: src/cpp/benders/benders_core/CriterionComputation.cpp [hunk 4]
```diff
@@ -29,19 +29,19 @@ void CriterionComputation::ComputeOuterLoopCriterion(
...
+const CriterionInputData &CriterionComputation::getOuterLoopInputData() const {
   return outer_loop_input_data_;
 }
 
 std::vector<std::vector<int>> &CriterionComputation::getVarIndices() {
   return var_indices_;
 }
 CriterionComputation::CriterionComputation(
-    const OuterLoopInputData &outer_loop_input_data)
+    const CriterionInputData &outer_loop_input_data)
```
Comment (tbittar) [inline src/cpp/benders/benders_core/CriterionComputation.cpp]:
```suggestion
    const CriterionInputData &criterion_input_data)
```

File: src/cpp/benders/benders_core/CriterionInputDataReader.cpp [hunk 5]
```diff
@@ -130,7 +127,7 @@ struct convert<OuterLoopSingleInputData> {
       err_msg << PrefixMessage(LogUtils::LOGLEVEL::FATAL, "Outer Loop")
               << "Error could not read 'area' field in outer loop input file"
               << "\n";
-      throw OuterLoopCouldNotReadAreaField(err_msg.str(), LOGLOCATION);
+      throw CriterionCouldNotReadAreaField(err_msg.str(), LOGLOCATION);
```
Comment (tbittar) [inline src/cpp/benders/benders_core/CriterionInputDataReader.cpp]:
```suggestion
      throw CouldNotReadAreaField(err_msg.str(), LOGLOCATION);
```

File: src/cpp/benders/benders_core/CriterionInputDataReader.cpp [hunk 6]
```diff
@@ -140,7 +137,7 @@ struct convert<OuterLoopSingleInputData> {
           << PrefixMessage(LogUtils::LOGLEVEL::FATAL, "Outer Loop")
           << "Error could not read 'criterion' field in outer loop input file"
           << "\n";
-      throw OuterLoopCouldNotReadCriterionField(err_msg.str(), LOGLOCATION);
+      throw CriterionCouldNotReadCriterionField(err_msg.str(), LOGLOCATION);
```
Comment (tbittar) [inline src/cpp/benders/benders_core/CriterionInputDataReader.cpp]:
```suggestion
      throw CouldNotReadCriterionField(err_msg.str(), LOGLOCATION);
```

File: src/cpp/benders/benders_core/VariablesGroup.cpp [hunk 7]
```diff
@@ -14,7 +14,7 @@ using namespace Benders::Criterion;
 
 VariablesGroup::VariablesGroup(
     const std::vector<std::string>& all_variables,
-    const std::vector<OuterLoopSingleInputData>& outer_loop_single_input_data)
+    const std::vector<CriterionSingleInputData>& outer_loop_single_input_data)
```
Comment (tbittar) [inline src/cpp/benders/benders_core/VariablesGroup.cpp]:
```suggestion
    const std::vector<CriterionSingleInputData>& criterion_single_input_data)
```

File: src/cpp/benders/benders_mpi/BendersMPI.cpp
Comment (tbittar) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp:1]: Rename `SubproblemData.outer_loop_criterions` and `SubproblemData.outer_loop_pattern_values` as `SubproblemData.criterions` and `SubproblemData.pattern_values` (L181-182)?

**Review by tbittar — CHANGES_REQUESTED** [review]

File: src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/CriterionComputation.h
Comment (a-zakir) [reply src/cpp/benders/benders_core/include/antares-xpansion/benders/benders_core/CriterionComputation.h:1]: done

File: src/cpp/benders/benders_mpi/BendersMPI.cpp
Comment (a-zakir) [reply src/cpp/benders/benders_mpi/BendersMPI.cpp:1]: yes good idea!

**Review by tbittar — APPROVED** [review]

**Pull request merged**